File: scaffolding/src/main/java/cn/bingoogolapple/scaffolding/view/MvpFragment.java
/**
 * Copyright 2016 bingoogolapple
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.bingoogolapple.scaffolding.view;

import android.support.annotation.StringRes;

import com.trello.rxlifecycle.LifecycleProvider;

import cn.bingoogolapple.scaffolding.presenter.BasePresenter;
import cn.bingoogolapple.scaffolding.util.ToastUtil;

/**
 * 作者:王浩 邮件:bingoogolapple@gmail.com
 * 创建时间:15/9/2 下午10:57
 * 描述:
 */
public abstract class MvpFragment<P extends BasePresenter> extends MvcFragment implements BaseView {
    protected P mPresenter;

    @Override
    public void showMsg(@StringRes int resId) {
        ToastUtil.show(resId);
    }

    @Override
    public void showMsg(String msg) {
        ToastUtil.show(msg);
    }

    @Override
    public MvcActivity getBaseActivity() {
        return mActivity;
    }

    @Override
    public LifecycleProvider getLifecycleProvider() {
        return this;
    }
}

File: scaffold/src/main/java/cn/bingoogolapple/scaffold/mvp/impl/MvpFragment.kt
/**
 * Copyright 2018 bingoogolapple
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.bingoogolapple.scaffold.mvp.impl

import android.os.Bundle
import cn.bingoogolapple.scaffold.mvp.IMvpView
import cn.bingoogolapple.scaffold.mvp.IPresenter
import cn.bingoogolapple.scaffold.view.MvcFragment
import java.lang.ref.WeakReference
import kotlin.coroutines.experimental.buildSequence
import kotlin.reflect.KClass
import kotlin.reflect.full.isSubclassOf
import kotlin.reflect.full.primaryConstructor
import kotlin.reflect.jvm.jvmErasure

/**
 * 作者:王浩 邮件:bingoogolapple@gmail.com
 * 创建时间:05/08/2018 7:16 PM
 * 描述:
 */
abstract class MvpFragment<out P : BasePresenter<MvpFragment<P>>> : IMvpView<P>, MvcFragment() {
    final override val mPresenter: P

    init {
        mPresenter = createPresenter()
        mPresenter.mView = WeakReference(this)
    }

    private fun createPresenter(): P {
        buildSequence {
            var thisClass: KClass<*> = this@MvpFragment::class
            while (true) {
                yield(thisClass.supertypes)
                thisClass = thisClass.supertypes.firstOrNull()?.jvmErasure ?: break
            }
        }.flatMap {
            it.flatMap { it.arguments }.asSequence()
        }.first {
            it.type?.jvmErasure?.isSubclassOf(IPresenter::class) ?: false
        }.let {
            return it.type!!.jvmErasure.primaryConstructor!!.call() as P
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mPresenter.onCreate(savedInstanceState)
    }

    override fun onStart() {
        super.onStart()
        mPresenter.onStart()
    }

    override fun onResume() {
        super.onResume()
        mPresenter.onResume()
    }

    override fun onPause() {
        super.onPause()
        mPresenter.onPause()
    }

    override fun onStop() {
        super.onStop()
        mPresenter.onStop()
    }

    override fun onDestroy() {
        mPresenter.onDestroy()
        super.onDestroy()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        mPresenter.onSaveInstanceState(outState)
    }

    override fun onViewStateRestored(savedInstanceState: Bundle?) {
        super.onViewStateRestored(savedInstanceState)
        mPresenter.onViewStateRestored(savedInstanceState)
    }
}